packages/salesforcedx-apex-debugger/src/adapter/apexDebug.ts

```diff
@@ -133,6 +141,102 @@ export class ApexDebug extends DebugSession {
+
+      try {
+        const linesToSetBreakpoint = await this.myBreakpointService.reconcileBreakpoints(
+          this.sfdxProject,
+          this.mySessionService.getSessionId(),
+          uriPath,
+          args.lines
+        );
+
+        for (const clientLine of linesToSetBreakpoint) {
+          const serverLine = this.convertClientLineToDebugger(clientLine);
+          const typeref = this.myBreakpointService.getTyperefFor(
+            uriPath,
+            serverLine
+          );
+          if (typeref && typeref.length > 0) {
+            const cmdOutput = await this.myBreakpointService.createLineBreakpoint(
+              this.sfdxProject,
+              this.mySessionService.getSessionId(),
+              typeref,
+              serverLine
+            );
+            this.myBreakpointService.cacheLiveBreakpoint(
+              uriPath,
+              clientLine,
+              cmdOutput.getId()
+            );
+            processedBreakpoints.push(
+              <DebugProtocol.Breakpoint>{
+                verified: true,
+                source: args.source,
+                line: clientLine
+              }
+            );
+          } else {
+            processedBreakpoints.push(
+              <DebugProtocol.Breakpoint>{
+                verified: false,
+                source: args.source,
+                line: clientLine
+              }
+            );
+          }
+        }
+
+        response.success = true;
+        response.body = {
+          breakpoints: processedBreakpoints
+        };
+      } catch (error) {
+        this.tryToParseSfdxError(response, error);
+      }
+    }
+
+    this.sendResponse(response);
+  }
+
+  protected customRequest(
+    command: string,
+    response: DebugProtocol.Response,
+    args: any
+  ): void {
+    switch (command) {
+      case 'lineBreakpointInfo':
+        const lineBpInfo: LineBreakpointInfo[] = args;
+        if (lineBpInfo && lineBpInfo.length > 0) {
+          const lineNumberMapping: Map<string, LineBpsInTyperef[]> = new Map();
+          for (const info of lineBpInfo) {
+            if (!lineNumberMapping.has(info.uri)) {
+              lineNumberMapping.set(info.uri, []);
+            }
+            const validLines: LineBpsInTyperef = {
+              typeref: info.typeref,
+              lines: info.lines
+            };
+            lineNumberMapping.get(info.uri)!.push(validLines);
+          }
+          this.myBreakpointService.setValidLines(lineNumberMapping);
+        }
+        // Make sure we have tried to query the language server
+        // before allowing VS Code to set breakpoints.
+        this.sendEvent(new InitializedEvent());
+        break;
+      default:
+        break;
+    }
+    this.sendResponse(response);
+  }
+
   private logToDebugConsole(msg?: string): void {
     if (msg && msg.length !== 0) {
       this.sendEvent(new OutputEvent(`${msg}${ApexDebug.TWO_NL}`));
```

Comment:
If you use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Multi-line_strings do you still need to use TWO_NL?

I'm not sure which one is cleaner but I thought I'd point it out in case you didn't know about it.